Fix part of #2269: Add query fronted page and handlers.

prasanna08 opened the pull request:
Added front-end page for submitting query and viewing result of past queries.
Added handlers for query data in back-end.

prasanna08 commented on core/domain/user_query_jobs_one_off.py:
```diff
@@ -102,3 +102,13 @@ def reduce(query_model_id, stringified_user_ids):
+
+    @classmethod
+    def _post_completed_hook(cls, job_id):
+        # TODO: change query status in query model to completed.
```
Hi @seanlip, this is actually part of milestone 4 (Implement success and failure) conditions but I have added them here for few difficulties I have been facing. Let me describe issue:

whenever a query has been completed, we have to change its status from processing to completed or failed. This can not be done inside `reduce()` function because of 2 reasons:
1. If no user is qualified for given query then reduce won't be called at all.
2. If query failed during map procedure reduce won't be called.

So I have to override _post_complete and _post_failure methods. To access query model from these functions we need query id, which is not available. Only job id is available. Now we need some way around this issue so that we can have query id in these methods. Then and only then completion and failure conditions can be performed. (That's why if you would run query from browser right now, it will never show completed status. It will always stay at processing status.)

Now to tackle this I can think of 2 ways.
1. We store additional params in job model for each job.
2. We store job id in query model along side query id.

So what do you think about this issue ? 
I added these methods here to ask this question and to tell that in this PR query will never go to complete status.

Thanks

seanlip replied:
Actually, if the additional params method works, I'm fine with that too. So, on the whole, I'm fine with both options -- whatever works.

seanlip replied:
OK, that's fine with me, if it works!

prasanna08 commented:
Hi @seanlip , added milestone 4 of #2269. 
I have added test for checking handler. Tell me if we have to add few more test. I can't think about other possible test cases.

Thanks.

seanlip commented on core/templates/dev/head/pages/query/Query.js:
```diff
@@ -0,0 +1,88 @@
+ * @fileoverview Data and controllers for the Oppia admin page.
+ */
+
+oppia.controller('Query', [
```
Let's call this the EmailDashboard.

pages/email_dashboard/EmailDashboard.js

seanlip commented on core/templates/dev/head/pages/query/Query.js:
```diff
@@ -0,0 +1,88 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Data and controllers for the Oppia admin page.
```
Wrong fileoverview.

seanlip commented on main.py:
```diff
@@ -374,6 +375,10 @@ def ui_access_wrapper(self, *args, **kwargs):
         r'%s/<collection_id>' % feconf.COLLECTION_RIGHTS_PREFIX,
         collection_editor.CollectionRightsHandler),
+
+    get_redirect_route(r'/query', user_query.QueryPage),
```
email_dashboard.EmailDashboardPage, etc.

('query' is too vague, it could mean a lot of things)

seanlip commented:
Thanks, @prasanna08! Took a very cursory preliminary pass -- can you fix these high-level things? I'll take another look tomorrow.

seanlip commented on core/controllers/user_query.py:
```diff
@@ -0,0 +1,124 @@
+
+    @require_valid_sender
+    def get(self):
+        offset = (
```
Instead of manually handling offsets, how about using GAE's built-in functionality for query cursors (e.g. with fetch_page())? This is also consistent with how we handle things like displaying search results, and avoids unnecessary bookkeeping.

https://cloud.google.com/appengine/docs/python/ndb/queries#cursors

prasanna08 replied:
I will check that.

Thanks

seanlip commented on core/controllers/user_query.py:
```diff
@@ -0,0 +1,124 @@
+            fetch(base_offset, offset=offset))
+        queries_list = []
+
+        for model in query_models:
```
You could also do:

```
queries_list = [{
    'id': model.id,
    'submitter_id': ...
    ...
} for model in query_models]
```

It's optional, but this is more consistent with how we tend to do it elsewhere.

seanlip commented on core/controllers/user_query.py:
```diff
@@ -0,0 +1,124 @@
+            query['id'] = model.id
+            query['submitter_id'] = user_services.get_username(
+                model.submitter_id)
+            query['created_on'] = model.created_on.strftime("%d-%m-%y %H:%M:%S")
```
Nit: use single quotes (in strftime arg)

seanlip commented on core/controllers/user_query.py:
```diff
@@ -0,0 +1,124 @@
+                model.submitter_id)
+            query['created_on'] = model.created_on.strftime("%d-%m-%y %H:%M:%S")
+            query['status'] = model.query_status
+            query['qualified_users'] = len(model.user_ids)
```
num_qualified_users or num_users, perhaps, otherwise the name suggests that this is a list.

seanlip commented on core/controllers/user_query.py:
```diff
@@ -0,0 +1,124 @@
+        """Post handler for query."""
+        data = self.payload['data']
+        kwargs = {}
+        if 'inactive_in_last_n_days' in data:
```
This is quite icky. Can we do a for loop?

seanlip commented on core/controllers/user_query_test.py:
```diff
@@ -0,0 +1,94 @@
+        self.process_and_flush_pending_tasks()
+
+    def test_that_page_is_accessible_to_authorized_users_only(self):
+        with self.assertRaisesRegexp(Exception, '401 Unauthorized'):
```
Put this only around the post_json call, so that we can be sure it's not due to anything else.

seanlip commented on core/templates/dev/head/pages/query/Query.js:
```diff
@@ -0,0 +1,88 @@
+
+oppia.controller('Query', [
+  '$scope', '$http', function($scope, $http) {
+    _QUERY_DATA_URL = '/querydatahandler';
```
Missing var. Also, we have stopped putting underscores in front of private functions, here and below. Instead, functions private to the controller are declared using var, and functions that need to be accessible from the template are declared on $scope.

seanlip commented on core/templates/dev/head/pages/query/Query.js:
```diff
@@ -0,0 +1,88 @@
+          _validate($scope.created_fewer_than_n_exps) &&
+          _validate($scope.edited_at_least_n_exps) &&
+          _validate($scope.edited_fewer_than_n_exps)) {
+        data = {
```
Missing var.

seanlip commented on core/controllers/user_query.py:
```diff
@@ -0,0 +1,124 @@
+        data = self.payload['data']
+        kwargs = {}
+        if 'inactive_in_last_n_days' in data:
+            if data['inactive_in_last_n_days'] != '':
```
Why empty string? I would expect null.

seanlip commented on core/templates/dev/head/pages/query/Query.js:
```diff
@@ -0,0 +1,88 @@
+          _validate($scope.edited_at_least_n_exps) &&
+          _validate($scope.edited_fewer_than_n_exps)) {
+        data = {
+          has_not_logged_in_for_n_days: $scope.has_not_logged_in_for_n_days,
```
Can these be numbers or None?

seanlip replied:
Actually, I just realized that there's a problem with the form, right? Once I've entered a value for some field, how do I remove/reset that value without refreshing the page? Setting it to something like 0 isn't always going to work...

Perhaps you might need some checkbox + number input, or a select2-like thing where you can "add tags" to a field...

With regards to the backend, we should validate/normalize all inputs before doing anything further with them. If that boundary is strong then we are less susceptible to errors in the frontend.

prasanna08 replied:
Or may be add a reset button (to set all the fields to None ?). We can have checkbox + input but reset sounds better to me. Still I don't understand what do you suggest by select2-like ?

seanlip replied:
Sort of like this, I was thinking: https://select2.github.io/examples.html#tags -- but I'm not sure how to handle the additional number parameter.

With reset, will it be clear when a field is empty versus when it has a value of zero? If so, I think that might be OK, since this page has a small audience; we can make it doubly-clear by putting a check mark on the right of the field if it's non-empty. Might want to style the form nicely, though, using ,form-control or similar -- see e.g. the input fields in the editor settings tab.

prasanna08 replied:
Yup. With reset it will be as if that parameter were never added in the first place. Totally blank query data.

seanlip commented on core/templates/dev/head/pages/query/query.html:
```diff
@@ -0,0 +1,87 @@
+  <div class="oppia-query-form" ng-controller="Query">
+    <p class="form-warning" ng-if="showWarning">*Please enter numeric data only.</p>
+    <form class="form">
+      Has not logged in for n days: <input type="text" ng-model="has_not_logged_in_for_n_days">
```
Suggest type="number", ditto below.

Force numbers to be non-negative.

seanlip commented on core/templates/dev/head/pages/query/Query.js:
```diff
@@ -0,0 +1,88 @@
+    var _fetchRecentQueries = function(offset) {
+      $http.get(_QUERY_DATA_URL, {
+        params: {
+          offset: offset,
```
Cursor would be better. (Also, reading this, I have no idea what the difference is between offset and base_offset...)

prasanna08 replied:
Oh, actually offset is offset and base_offset is like num_of_recent_query_per_page. Something like that. I just could not think of a better (and short) name for it.

prasanna08 replied:
May be fix (constant) it in back end it self. Not passing it at all.
I am still figuring out to make this work with cursors (that link you suggested). Once that is implemented, this whole logic is going to be changed (offset, base_offset etc..). Once I am done with that you can review it and tell me more suggestions. Does that sound good ?

seanlip replied:
Yep, sgtm (on all counts). Thanks!

seanlip commented on core/templates/dev/head/pages/query/Query.js:
```diff
@@ -0,0 +1,88 @@
+        var data = response.data;
+        if (data.recent_queries.length > 0) {
+          $scope.recentQueries = data.recent_queries;
+          $scope.displayRecentQueries = data.recent_queries.length;
```
Odd variable name; it suggests a function since it starts with a verb.

Not sure you need it, though. Could just reference recentQueries.length from the template...

seanlip commented on core/templates/dev/head/pages/query/Query.js:
```diff
@@ -0,0 +1,88 @@
+      $scope.showSuccessMessage = false;
+    };
+
+    $scope.nextRecentQueries = function() {
```
Function names should start with verbs.

seanlip commented on core/templates/dev/head/pages/query/Query.js:
```diff
@@ -0,0 +1,88 @@
+      _fetchRecentQueries(offset);
+    };
+
+    $scope.previousRecentQueries = function() {
```
Why not cache in frontend? Maintain the full list here, plus a cursor to indicate where to continue the search from if more entries need to be fetched. Also maintain a separate 'offset' var in the frontend to denote what range of queries the user is currently looking at.

prasanna08 replied:
That could work too. It would improve performance but have more memory usage.

seanlip commented on core/templates/dev/head/pages/query/query.html:
```diff
@@ -0,0 +1,87 @@
+  <ul class="nav navbar-nav oppia-navbar-breadcrumb">
+    <li>
+      <span class="oppia-navbar-breadcrumb-separator"></span>
+      Query
```
Email Dashboard

seanlip commented on core/controllers/user_query.py:
```diff
@@ -0,0 +1,124 @@
+        for model in query_models:
+            query = {}
+            query['id'] = model.id
+            query['submitter_id'] = user_services.get_username(
```
I think you might need to optimize this using get_multi(). There could be a lot of entries in this list.

See core.domain.summary_services.get_library_groups() for an example where we have to deal with a similar issue.

prasanna08 replied:
Oh, I will check that.

Thanks

prasanna08 commented:
Hi @seanlip added new commit attending all of your suggestions.
Changes:
- Cached query results in fronted.
- Used cursor instead of manually handling offset.
- Used get_multi to fetch all submitter names at once.
- Added reset button to the form.
- Stored additional query params in job model so that now query status can be changed to completed or failed.
- Changed input type of form to number so now we can pass received data directly to the query model without any checking. Its values will be either a number or None (which means parameter should not be considered as part of query). Type checking is handled by html5 itself. So no need of validate function either.

That's all. Tell me if I forgot something.
Thanks

seanlip commented on core/controllers/email_dashboard.py:
```diff
@@ -14,6 +14,8 @@
 
 """Controller for user query related pages and handlers."""
 
+from google.appengine.datastore.datastore_query import Cursor
```
Confine imports of GAE-related stuff to gae_models and core/platform -- i.e., pass the urlsafe string through and convert it to an actual cursor only in the storage layer. Define an additional helper method in gae_models if necessary.

(This keeps the GAE dependency nicely separated from the rest of the codebase.)

seanlip commented on core/controllers/email_dashboard.py:
```diff
@@ -41,79 +43,59 @@ def test_user(self, **kwargs):
+            fetch_page(self.QUERIES_PER_PAGE, start_cursor=cursor))
+
+        submitters_settings = user_services.get_users_settings(
+            [model.submitter_id for model in query_models])
```
Do the elements in this list need to be made unique? I'm not sure.

seanlip commented on core/controllers/email_dashboard.py:
```diff
@@ -41,79 +43,59 @@ def test_user(self, **kwargs):
+        data = {
             'recent_queries': queries_list
-        })
+        }
```
Add a line here:

     'cursor': next_cursor.urlsafe() if (more and next_cursor) else None

That way, we always return a fixed set of keys.

seanlip commented on core/controllers/email_dashboard.py:
```diff
@@ -41,79 +43,59 @@ def test_user(self, **kwargs):
     @require_valid_sender
     def post(self):
         """Post handler for query."""
         data = self.payload['data']
```
'data' should be validated, and there should be tests to ensure that a 400 error is raised if invalid data is passed through.

seanlip commented on core/controllers/email_dashboard_test.py:
```diff
@@ -79,16 +79,9 @@ def test_that_handler_works_correctly(self):
-                    }
-                }, csrf_token)
+            self.testapp.get('/emaildashboard')
             self.logout()
```
Put this outside the 'with'.

seanlip commented on core/domain/user_query_jobs_one_off.py:
```diff
@@ -105,10 +107,18 @@ def reduce(query_model_id, stringified_user_ids):
     def _post_completed_hook(cls, job_id):
-        # TODO: change query status in query model to completed.
-        pass
+        # TODO: send email to submitter upon completion.
```
TODO(prasanna08): ...

seanlip commented on core/domain/user_query_jobs_one_off.py:
```diff
@@ -105,10 +107,18 @@ def reduce(query_model_id, stringified_user_ids):
     def _post_failure_hook(cls, job_id):
-        # TODO: change query status in query model to failed.
-        pass
+        # TODO: send email to admin and submitter upon failure.
```
Ditto.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -0,0 +1,121 @@
+    var QUERY_DATA_URL = '/emaildashboarddatahandler';
+
+    // Store current cursor.
+    $scope.currentCursor = null;
```
Hm, this is an interesting approach, but I want to suggest a different one that I think is better design-wise and that is more consistent with what we do in other places where we have paging.

In general, we treat cursors as only having relevance in the backend, and don't want to spread them all over the frontend codebase. So, this controller should maintain only the latest cursor, and nothing else.

In addition, the controller should have an array along the lines of $scope.queries, a constant $scope.PAGE_SIZE, and a variable $scope.currentPageIndex.

Initially, $scope.queries = [], $scope.PAGE_SIZE = 10, $scope.currentPageIndex = 0. We make our first fetch, and $scope.queries has (up to) 10 entries. If the backend returns a cursor we display a 'next' arrow/button, otherwise we disable it.

When the 'next'/'previous' arrows/buttons are clicked, $scope.currentPageIndex increments/decrements by 1, as needed. The exception is when the 'next' arrow is clicked and we're at the end of the $scope.queries array, in which case we make another fetch with our current cursor to add more entries to $scope.queries, and then go forward in the array. In addition, when a new query is created, it should be pushed into the front of the array (after the backend has accepted it).

That way, you have an array $scope.queries and an index into it, which I think is a better data structure to represent what's on the page. You might also find it useful to have a look at core/templates/dev/head/pages/profile/Profile.js for an example that implements the frontend part of this approach. Let me know if you have any questions about this!

prasanna08 replied:
You suggest that we keep only latest cursor and all the rest of queries should be cached but what about Reload button. If we use such data structure user won't be able to use Reload button. Reloading any set of queries from backend will require cursor of that set.

So if we need Reload button then we have to store cursor + queries. In addition to that we will also need next_cursor obtained during last fetch. So in that case we need some thing like repeated dictionaries.
```
$scope.queries = [ {
    queries: (queries of particular page)
    cursor: (cursor for reloading particular page)
 } for each page fetched from backend]
```

Your approach works if reload button is not present. I am fine with either way. We completely remove Reload button or we use something like above to keep it.

seanlip replied:
Ah, for that case, how about having a "re-check status" button for the individual job? If I'm understanding correctly, it seems a bit wasteful to reload the entire page...

prasanna08 replied:
Oh, re-check status sounds good but in that case we will need entire new handler. Which will take input 'query_id' and then send latest info from query model. Since queries are listed from a list, entry of this query in the list will have to be updated (in both $scope.queries and $scope.recentQueries). I think this can be done.

seanlip replied:
Yep, that sounds good to me.

One other thought: I suggest that you create an Angular service to store the data, and keep the controller minimal. This is good practice because a service can be more easily tested, and you won't end up mixing all the UI stuff with the "model" layer and logic. We're planning to move gradually towards this structure in all our frontend pages that have significant logic components, in order to make the code easier to maintain going forward.

Here's an [example](https://github.com/oppia/oppia/blob/simple-editor-2/core/templates/dev/head/pages/exploration_editor/simple_editor_tab/SimpleEditorManagerService.js) that shows how I did this recently for a simpler version of the editor that we're working on. Note that `data` is a private variable that maintains the "state" of the page, and the service exposes a bunch of helper methods like getTitle(), saveTitle(), addNewQuestion() etc. You could do something similar here. In addition, it would also be good to have a Query with a setStatus() method, similar to how we have Question objects [here](https://github.com/oppia/oppia/blob/simple-editor-2/core/templates/dev/head/domain/exploration/QuestionObjectFactory.js) -- you can put that in domain/email_dashboard.

(I realize that this is a bit more work at the outset, but I think it would be a good idea to try and practice making the frontend clean and maintainable from the beginning.)

prasanna08 commented:
Hi @seanlip added new commit with following change:
- Designed new frontend logic as we discussed.
- Moved gae-cursor related functions to a new platform file.
- Added data validation check when submitting query (400 InvalidInputException) and added backend tests.

Thanks. Tell me if I am missing something.

seanlip commented on core/controllers/email_dashboard.py:
```diff
@@ -0,0 +1,142 @@
+from core.controllers import base
+from core.domain import config_domain
+from core.domain import user_query_services
+from core.domain import user_query_jobs_one_off
```
Alphabetize.

seanlip commented on core/controllers/email_dashboard.py:
```diff
@@ -0,0 +1,142 @@
+(user_models,) = models.Registry.import_models([models.NAMES.user])
+
+current_user_services = models.Registry.import_current_user_services()
+cursor_services = models.Registry.import_cursor_services()
```
Do we really need this new abstraction, though? Would it be possible, instead, to pass the urlsafe cursor into a new method in the gae_models class which acts as a seam to fetch_page() -- and pass a urlsafe cursor out from that method as well? That would still preserve the separation, but reduce the amount of code "surface" that needs to be maintained.

prasanna08 replied:
So you suggest to create a new service for email dashboard in domain which will do fetch page and return fetched queries and next_cursor.urlsafe(). Is that correct ?

seanlip replied:
No -- I suggest creating a method like user_models.UserQueryModel.fetch_page(), and calling that from here with a urlsafe cursor. No need to introduce new files.

prasanna08 replied:
Okay, understood that.

Thanks

seanlip commented on core/controllers/email_dashboard.py:
```diff
@@ -0,0 +1,142 @@
+            'edited_at_least_n_exps', 'edited_fewer_than_n_exps']
+
+        for key, value in data.iteritems():
+            if key not in possible_keys or not isinstance(value, int):
```
Check that the int is >= 0.

seanlip commented on core/platform/cursor/gae_cursor_services.py:
```diff
@@ -0,0 +1,23 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Provides app identity services."""
```
Docstring is wrong. But I don't think we need this file -- see above.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -0,0 +1,168 @@
+    var QUERIES_PER_PAGE = 10;
+    // Store latest cursor value for fetching next query page.
+    var latestCursor = null;
+    // Cache all fetched queries.
```
Array containing all fetched queries.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -0,0 +1,168 @@
+    var latestCursor = null;
+    // Cache all fetched queries.
+    var queries = [];
+    // Index of currently showing page of query results.
```
currently-shown

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -0,0 +1,168 @@
+ * @fileoverview Controller and factory for oppia email dashboard page.
+ */
+
+oppia.factory('EmailDashboardDataService', [
```
This should go in a new file. One service/controller/directive/etc. per file.

seanlip replied:
Also, please add some Karma unit tests for this service.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -0,0 +1,168 @@
+
+        var startQueryIndex = (currentPageIndex) * QUERIES_PER_PAGE;
+        var endQueryIndex = (currentPageIndex + 1) * QUERIES_PER_PAGE;
+        fetchQueries().then(function(data) {
```
Wouldn't the cursor be incorrect?

Also, do you need to fetch from the backend again? You know all the query details, so you could just stick the new query in the beginning of the list -- and if the $http.post() fails, remove it again.

prasanna08 replied:
No, we do not know the query_id of query, time of submission etc.

seanlip replied:
Ah ok, fair enough. But then you don't need to update the cursor when you do this fetch, I think?

Also is it possible to just get the data for the first entry, rather than the whole page? Maybe using a parameter to say "here is the number of queries I want"?

prasanna08 replied:
Okay, I think we can customise that new handler I added (for rechecking query status). If it has `query_id` present in request we will give that query or else it should have `num_of_queries` in request so we will return that much queries. How does that sound ?

seanlip replied:
I was actually thinking, why don't we always send the number of queries that the frontend wants when making a fetch request? The backend should check that the number of queries is <= 50 (say) to prevent server overload, but other than that it seems reasonable to let the client be in control of it...

prasanna08 replied:
I don't understand your question ?
We are using fetch_page() method and  we show 10 queries per page. That value is fixed in backend. So client is not controlling it, server is controlling how many queries to display per page.
So why do we need to send number of queries to fetch to backend ?

seanlip replied:
Yes, exactly. I was wondering why we need the server to control it at all, since the notion of a page is really entirely in the frontend. Perhaps the frontend should just pass it in?

prasanna08 replied:
So what do you suggest ? We have another field on frontend so user can enter num_of_queries_per_page and then send this to backend as page size ?

prasanna08 replied:
Ah, I see.
 That could work.
Thanks

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -0,0 +1,168 @@
+        var endQueryIndex = (currentPageIndex + 2) * QUERIES_PER_PAGE;
+        currentPageIndex = currentPageIndex + 1;
+
+        if (queries.length > startQueryIndex) {
```
I think you need to do this a bit more carefully. If you fetch two pages and then add a query to the beginning of the list, the query that was at index 30 is now going to be at index 31.

Instead, when the 'next' button is pressed, do a fetch only if endQueryIndex has not been reached and if there is still a cursor (indicating that there is still more stuff to fetch).

prasanna08 replied:
Ah, yes that would not be correct. 

Thanks

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -0,0 +1,168 @@
+      $scope.showSuccessMessage = true;
+    };
+
+    $scope.getNextRecentQueries = function() {
```
getNextPageOfQueries

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -0,0 +1,168 @@
+      }
+    };
+
+    $scope.getPreviousRecentQueries = function() {
```
getPreviousPageOfQueries

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -0,0 +1,168 @@
+oppia.controller('EmailDashboard', [
+  '$scope', '$http', 'EmailDashboardDataService',
+  function($scope, $http, EmailDashboardDataService) {
+    $scope.recentQueries = [];
```
$scope.currentPageOfQueries

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard.html:
```diff
@@ -0,0 +1,85 @@
+          </tr>
+        </thead>
+        <tbody>
+          <tr ng-repeat="q in recentQueries">
```
add "track by $index"?

prasanna08 replied:
Ah yes forgot that.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -0,0 +1,168 @@
+            query_id: query.id
+          }
+        }).then(function(response) {
+          var data = response.data;
```
Shouldn't this update the `queries` var, too?

prasanna08 replied:
Yup missed that.

Thanks

prasanna08 commented:
Hi @seanlip added new commit addressing all your previous comments.
Changes:
- Separated email dashboard service and controller in two separate files.
- Added karma test for email dashboard.
- Removed cursor files from platform and added fetch_page() method in query data store model.
- Other few changes (As mentioned in comments).

Thanks

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -0,0 +1,81 @@
+      $scope.currentPageOfQueries = queries;
+    };
+
+    var setQueryStatus = function(index, query) {
```
This is more like setQuery

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -0,0 +1,81 @@
+
+    $scope.getNextPageOfQueries = function() {
+      if (EmailDashboardDataService.isNextPageAvailable()) {
+        EmailDashboardDataService.getNextQueries(updateCurrentPageOfQueries);
```
Can you do this as a promise, instead?
```
EmailDashboardDataService.getNextQueries().then(updateCurrentPageOfQueries);
```
or
```
EmailDashboardDataService.getNextQueries().then(function(queries) {
  updateCurrentPageOfQueries(queries);
});
```

I think it reads better that way. Ditto below.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -0,0 +1,81 @@
+ */
+
+oppia.controller('EmailDashboard', [
+  '$scope', '$http', 'EmailDashboardDataService',
```
No need $http?

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardServices.js:
```diff
@@ -0,0 +1,120 @@
+    var fetchQueriesPage = function(pageSize, cursor) {
+      return $http.get(QUERY_DATA_URL, {
+        params: {
+          num_of_queries_per_page: pageSize,
```
num_queries_to_fetch

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardServices.js:
```diff
@@ -0,0 +1,120 @@
+        $http.post(QUERY_DATA_URL, {
+          data: data
+        }).success(function() {
+          // Wait for 500ms and then fetch details of newly submitted query.
```
Why don't you just get the POST to return this data, instead of doing an additional fetch?

prasanna08 replied:
Good idea.

Thanks

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardServices.js:
```diff
@@ -0,0 +1,120 @@
+        var endQueryIndex = (currentPageIndex + 2) * QUERIES_PER_PAGE;
+        currentPageIndex = currentPageIndex + 1;
+
+        if (queries.length > endQueryIndex) {
```
Should this be >= ?

prasanna08 replied:
Oh, yes. You are right.

Thanks

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardServices.js:
```diff
@@ -0,0 +1,120 @@
+
+      isNextPageAvailable: function() {
+        var nextQueryIndex = (currentPageIndex + 1) * QUERIES_PER_PAGE;
+        return !(queries.length <= nextQueryIndex) || Boolean(latestCursor);
```
Should the first clause just be queries.length >= nextQueryIndex? (no need for the negation)

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardServices.js:
```diff
@@ -0,0 +1,120 @@
+        return (currentPageIndex > 0);
+      },
+
+      recheckQueryStatus: function(index, callback) {
```
Based on what this method is doing, it seems more like "fetchQuery".

seanlip replied:
Any reason not to just pass the query id directly?

prasanna08 replied:
I passed index because without index it would have to iterate over whole queries list but as you suggested in comment below (to iterate over queries since index might not be stable), I agree with it and changing function's working that way won't require index.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardServices.js:
```diff
@@ -0,0 +1,120 @@
+      },
+
+      recheckQueryStatus: function(index, callback) {
+        var queryIndex = (currentPageIndex) * QUERIES_PER_PAGE + index;
```
Don't need the parentheses.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardServices.js:
```diff
@@ -0,0 +1,120 @@
+          }
+        }).then(function(response) {
+          var data = response.data;
+          queries[queryIndex] = data.query;
```
Hm, the index may not be stable (e.g. a new query may be inserted into the front of the list in the meantime). I suggest iterating over the queries until you find the one with the same query id and then replacing that; this way, you can be confident that you're replacing the correct thing. We've run into problems like this for e.g. multiple choice question options.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardServicesSpec.js:
```diff
@@ -0,0 +1,54 @@
+ * @fileoverview Unit tests for the email dashboard page.
+ */
+
+describe('Email Dashboard Services', function() {
```
Add tests to simulate, e.g., let's say you have 25 queries, try paging forward two times, page backwards once, add a query, page forward again -- make sure the page numbers are correct and that you end up with 26 queries. Also, when fetching a query make sure that the correct one is replaced.

(You'll need to break this down into individual tests, but the upshot is that there ought to be more tests here.)

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard.html:
```diff
@@ -0,0 +1,86 @@
+{% extends 'pages/base.html' %}
+
+{% block maintitle %}
+  Query - Oppia
```
Email Dashboard - Oppia

prasanna08 commented on core/storage/user/gae_models.py:
```diff
@@ -362,3 +363,12 @@ class UserQueryModel(base_models.BaseModel):
+
+    @classmethod
+    def fetch_page(cls, page_size, cursor):
+        cursor = Cursor(urlsafe=cursor)
```
Yes, when cursor is None it will fetch very first page of queries.

prasanna08 commented:
Hi @seanlip added new commit with following major changes:
- Added new tests for email dashboard data services as you suggested.
- Removed all callbacks and implemented controller based on promise functionality.
- Also modified post handler to also return data of the newly submitted query.

Thanks

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardServices.js:
```diff
@@ -53,66 +53,70 @@ oppia.factory('EmailDashboardDataService', [
-            });
-          }, 500);
+        }).then(function(response) {
+          var data = response.data;
```
Optional -- do you think, after a submitQuery(), the page number should reset to the first page? I'm not sure, and I'm fine either way -- your call.

prasanna08 replied:
I really don't know which one would be better. I think we should let it be the way it is right now since it shows a message about query being submitted successfully after submitting it. We can change it later on if required.

seanlip commented on core/controllers/email_dashboard.py:
```diff
@@ -119,12 +133,13 @@ def validate(self, data):
                     value < 0):
                 # Raise exception if key is not one of the allowed keys or
-                # correspoding value is not of type integer.
+                # corresponding value is not of type integer..
```
Duplicate full stop at end...
